File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/ChainOfResponsibility/handler.go
package ChainOfResponsibility

import "fmt"

// 日志级别常量
const (
	DEFAULT = 1
	WARNING = 2
	ERROR   = 3
)

func defaultWrite(message string) {
	fmt.Println("Default Console: " + message)
}

type BaseLogger struct {
	l          LoggerHandler
	level      int
	nextLogger *LoggerHandler
}

// LoggerHandler 日志处理器接口
type LoggerHandler interface {
	Log(level int, message string)
	Write(message string)
	GetNextLogger() *LoggerHandler
}

func SetNextLogger(logger LoggerHandler, nextLogger LoggerHandler) {
	*logger.GetNextLogger() = nextLogger
}

func (h *BaseLogger) Log(level int, message string) {
	if level == h.level {
		// 调用具体实现的 Write 方法
		if logger, ok := interface{}(h).(LoggerHandler); ok {
			logger.Write(message)
		}
		return
	}
	if h.nextLogger != nil {
		(*h.nextLogger).Log(level, message)
	} else {
		defaultWrite(message)
	}
}

func (h *BaseLogger) GetNextLogger() *LoggerHandler {
	return h.nextLogger
}

// Write 写入日志（需要子类实现）
func (h *BaseLogger) Write(message string) { h.l.Write(message) }

// NewBaseLogger 创建基础日志处理器
func NewBaseLogger(level int, logger LoggerHandler) *BaseLogger {
	return &BaseLogger{
		l:          logger,
		level:      level,
		nextLogger: new(LoggerHandler),
	}
}

File: DesignPatterns/01_GOF23/DP_examples/Rust/03_Behavioral/ChainOfResponsibility/handler.rs
// 日志级别常量
pub const DEFAULT: i32 = 1;
pub const WARNING: i32 = 2;
pub const ERROR: i32 = 3;

// 日志处理器 trait
pub trait LoggerHandler {
    // 设置下一个处理器
    fn set_next_logger(&mut self, logger: Box<dyn LoggerHandler>);
    // 记录日志
    fn log(&self, level: i32, message: &str);
    // 写入日志（由具体实现）
    fn write(&self, message: &str);
}

// 基础日志处理器实现
pub struct BaseLoggerHandler {
    level: i32,
    next_logger: Option<Box<dyn LoggerHandler>>,
}

impl BaseLoggerHandler {
    // 创建基础日志处理器
    pub fn new(level: i32) -> Self {
        BaseLoggerHandler {
            level,
            next_logger: None,
        }
    }
}

impl LoggerHandler for BaseLoggerHandler {
    // 设置下一个处理器
    fn set_next_logger(&mut self, logger: Box<dyn LoggerHandler>) {
        self.next_logger = Some(logger);
    }

    // 记录日志
    fn log(&self, level: i32, message: &str) {
        if level == self.level {
            self.write(message);
        } else if let Some(ref next) = self.next_logger {
            next.log(level, message);
        } else {
            // 默认处理
            println!("Default Console: {}", message);
        }
    }

    // 写入日志（默认实现，由子类重写）
    fn write(&self, _message: &str) {
        // 由具体实现重写
    }
}
